### PR TITLE
[Bug-fix]: Migration pod becomes "Evicted" since no ephemeral-storage is requested

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
This PR fixes the following bug: https://bugzilla.redhat.com/show_bug.cgi?id=1731819

Migration pods became evicted since no they used ephemeral-storage that they didn't request. Therefore this PR adds an ephemeral-storage of small amount (1MB) to both compute and volume-disk containers.

**Special notes for your reviewer**:
Unfortunately I couldn't reproduce the error. I've consulted @vladikr which said it might be very difficult to reproduce the bug locally.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/patches.go [hunk 1]
```diff
@@ -181,14 +181,22 @@ func (c *Customizer) GetPatchesForResource(resourceType, name string) []v1.Custo
 	patches := make([]v1.CustomizeComponentsPatch, 0)
 
 	for _, p := range allPatches {
-		if strings.EqualFold(p.ResourceType, resourceType) && strings.EqualFold(p.ResourceName, name) {
+		if valueMatchesKey(p.ResourceType, resourceType) && valueMatchesKey(p.ResourceName, name) {
```
Comment (vladikr) [inline pkg/virt-operator/resource/apply/patches.go:184]: @iholder-redhat do you need this change to be part of this PR?

File: api/openapi-spec/swagger.json [hunk 2]
```diff
@@ -9242,6 +9242,12 @@
    },
    "v1.CustomizeComponentsPatch": {
     "type": "object",
+    "required": [
+     "resourceName",
+     "resourceType",
+     "patch",
+     "type"
+    ],
```
Comment (vladikr) [inline api/openapi-spec/swagger.json:9250]: Is this related to this PR? let's just keep only those changes that are relevant to this PR.

**Comment by iholder-redhat** [comment]
@vladikr I'm sorry, something went wrong here.. PR included commits I didn't intend it to.
I've re-pushed. Now it looks fine.

File: pkg/virt-controller/services/template.go [hunk 3]
```diff
@@ -713,6 +713,10 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
 		resources.Requests[key] = value
 	}
 
+	if _, ephemeralStorageDefined := resources.Requests[k8sv1.ResourceEphemeralStorage]; !ephemeralStorageDefined {
+		resources.Requests[k8sv1.ResourceEphemeralStorage] = resource.MustParse("1Mi")
```
Comment (iholder-redhat) [inline pkg/virt-controller/services/template.go:717]: @stu-gott raised a good point here: if the user requests X MBs of ephemeral-storage then the amount that should be requested is X+1Mb since we should guarantee the user can all of his requested storage.

I'm updating and pushing again.

Comment (vladikr) [reply pkg/virt-controller/services/template.go:717]: users cannot request anything for container-disk, only for the VMI. so this overhead should be explicitly set for the container-disk.
For the compute container, I don't think we've ever considered users requesting this, but we don't prevent it.
I think you can follow the example of how memory overhead is being added and do something similar with the ephemeral-storage overhead 
https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/services/template.go#L777-L786

**Review by davidvossel — CHANGES_REQUESTED** [review]
this shoudn't be possible.  set evictionStrategy: LiveMigrate on the VMI spec and evictions will be blocked.

File: pkg/virt-controller/services/template.go (hunk 3 above)
Comment (davidvossel) [reply pkg/virt-controller/services/template.go:717]: I think we should increase this to 50Mi to account for unknowns here. storage is cheap, and the repercussions for getting this wrong are high.

**Comment by davidvossel** [comment]
> this shoudn't be possible. set evictionStrategy: LiveMigrate on the VMI spec and evictions will be blocked.

ignore me. I am wrong here. This eviction is being caused by the kubelet due to disk pressure. If we're consuming any ephemeral storage, then we'll be over our request of (0Mi), which makes us eligible for termination. 

i think this pr is going in the right direction

File: pkg/virt-controller/services/template.go (hunk 3 above)
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:717]: Thansk @davidvossel . Do you mean 50M to both compute and container disk?

Comment (vladikr) [reply pkg/virt-controller/services/template.go]: > Thansk @davidvossel . Do you mean 50M to both compute and container disk?

Yes, 50Mi for both should be fine.

Comment (vladikr) [reply pkg/virt-controller/services/template.go]: > Thansk @davidvossel . Do you mean 50M to both compute and container disk?

Yes, 50Mi for both should be fine.

File: pkg/container-disk/container-disk_test.go [hunk 4]
```diff
@@ -181,7 +181,29 @@ var _ = Describe("ContainerDisk", func() {
...
+					Expect(containerResourceSpec).To(And(HaveKey(k8sv1.ResourceCPU), HaveKey(k8sv1.ResourceMemory)))
+				}
+			})
+			It("by verifying that ephemeral storage request is set to every container", func() {
+
+				vmi := v1.NewMinimalVMI("fake-vmi")
+				appendContainerDisk(vmi, "r0")
+				containers := GenerateContainers(vmi, "libvirt-runtime", "/var/run/libvirt")
+
+				expectedEphemeralStorageRequest := resource.MustParse("50M")
```
Comment (vladikr) [inline pkg/container-disk/container-disk_test.go:197]: 50Mi ? not M, let's keep it consistent.

File: pkg/container-disk/container-disk.go [hunk 5]
```diff
@@ -179,6 +179,14 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, bi
 func generateContainersHelper(vmi *v1.VirtualMachineInstance, podVolumeName string, binVolumeName string, isInit bool) []kubev1.Container {
 	var containers []kubev1.Container
 
+	const (
+		Size1M   = "1M"
+		Size10M  = "10M"
+		Size50M  = "50M"
+		Size40M  = "40M"
+		Size100M = "100M"
```
Comment (vladikr) [inline pkg/container-disk/container-disk.go:187]: This actually makes it look like it's all the same units, while 10m and 100m you've replaced for CPU values, are in fact `milliCPU` units and memory is measured mebibytes, I guess it all just invites confusion.

Comment (vladikr) [reply pkg/container-disk/container-disk.go:187]: I would either avoid doing this or use explicit units.

File: pkg/virt-controller/services/template.go [hunk 6]
```diff
@@ -718,6 +718,21 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
 		resources.Limits[key] = value
 	}
 
+	// Add ephemeral storage request to container to be used by Kubevirt. This amount of ephemeral storage
+	// should be added to the user's request.
+	ephemeralStorageAddition := resource.MustParse("50M")
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go:723]: maybe ephemeralStorageOverhead? would be consistent with memory...

File: pkg/virt-controller/services/template.go [hunk 7]
```diff
@@ -718,6 +718,21 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
...
+	if ephemeralStorageRequested, ephemeralStorageDefined := resources.Requests[k8sv1.ResourceEphemeralStorage]; ephemeralStorageDefined {
+		ephemeralStorageRequested.Add(ephemeralStorageAddition)
+		resources.Requests[k8sv1.ResourceEphemeralStorage] = ephemeralStorageRequested
+	} else {
+		resources.Requests[k8sv1.ResourceEphemeralStorage] = ephemeralStorageAddition
+	}
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go:729]: I think this can simply be:
```
ephemeralStorageRequested := resources.Requests[k8sv1.ResourceEphemeralStorage]
ephemeralStorageRequested.Add(ephemeralStorageAddition)
resources.Requests[k8sv1.ResourceEphemeralStorage] = ephemeralStorageRequested
```

if the resources.Requests map doesn't have this key it will return just resource.Quantity{} and then you can simply add your overhead to it...

File: pkg/virt-controller/services/template.go [hunk 8]
```diff
@@ -718,6 +718,21 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
...
+
+	if ephemeralStorageLimit, ephemeralStorageLimitDefined := resources.Limits[k8sv1.ResourceEphemeralStorage]; ephemeralStorageLimitDefined {
+		ephemeralStorageLimit.Add(ephemeralStorageAddition)
+		resources.Limits[k8sv1.ResourceEphemeralStorage] = ephemeralStorageLimit
+	}
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go:734]: that's fine. we only want to add a limit if one was defined.

File: pkg/virt-controller/services/template.go (hunk 7 above)
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:729]: Cool!

File: pkg/virt-controller/services/template.go [hunk 9]
```diff
@@ -718,6 +718,18 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
 		resources.Limits[key] = value
 	}
 
+	// Add ephemeral storage request to container to be used by Kubevirt. This amount of ephemeral storage
+	// should be added to the user's request.
+	ephemeralStorageOverhead := resource.MustParse("50M")
```
Comment (enp0s3) [inline pkg/virt-controller/services/template.go:723]: it will become redundant if you will use a const var for `50M`

File: pkg/container-disk/container-disk.go [hunk 10]
```diff
@@ -187,22 +187,18 @@ func generateContainersHelper(vmi *v1.VirtualMachineInstance, podVolumeName stri
 			diskContainerName := fmt.Sprintf("volume%s", volume.Name)
 			diskContainerImage := volume.ContainerDisk.Image
 			resources := kubev1.ResourceRequirements{}
+			resources.Limits = make(kubev1.ResourceList)
+			resources.Requests = make(kubev1.ResourceList)
+			resources.Limits[kubev1.ResourceMemory] = resource.MustParse("40M")
+			resources.Requests[kubev1.ResourceCPU] = resource.MustParse("10M")
+			resources.Requests[kubev1.ResourceEphemeralStorage] = resource.MustParse("50M")
```
Comment (enp0s3) [inline pkg/container-disk/container-disk.go:194]: Should we add resource limits for the ephemeral-storage as well?

File: pkg/virt-controller/services/template_test.go [hunk 11]
```diff
@@ -2573,6 +2573,55 @@ var _ = Describe("Template", func() {
...
+
+			table.DescribeTable("by verifying that ephemeral storage ", func(defineEphemeralStorageLimit bool) {
+				vmi := v1.NewMinimalVMI("fake-vmi")
+
+				ephemeralStorageRequests := resource.MustParse("30M")
+				ephemeralStorageLimit := resource.MustParse("70M")
+				ephemeralStorageAddition := resource.MustParse("50M")
+
+				if defineEphemeralStorageLimit {
+					vmi.Spec.Domain.Resources = v1.ResourceRequirements{
```
Comment (enp0s3) [inline pkg/virt-controller/services/template_test.go:2586]: Sorry but I can't see how does the VMI resource request affects the container disk resource request, aren't the container disk resource request is hard-coded inside the `GenerateContainers` function?

**Review by enp0s3 — CHANGES_REQUESTED** [review]
Thank you! looks good, please see my comments below

File: pkg/container-disk/container-disk_test.go [hunk 12]
```diff
@@ -181,7 +181,29 @@ var _ = Describe("ContainerDisk", func() {
 					},
 				}
 				containers := GenerateContainers(vmi, "libvirt-runtime", "/var/run/libvirt")
-				Expect(containers[0].Resources.Limits).To(HaveLen(2))
+
+				containerResourceSpecs := []k8sv1.ResourceList{containers[0].Resources.Limits, containers[0].Resources.Requests}
+
+				for _, containerResourceSpec := range containerResourceSpecs {
+					Expect(containerResourceSpec).To(And(HaveKey(k8sv1.ResourceCPU), HaveKey(k8sv1.ResourceMemory)))
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/container-disk/container-disk_test.go:188]: This is a test I did not write. The test verifies that the resources that were explicitly defined in VMI (in this case memory and CPU request/limit) are indeed defined in container specs. The original test implementation expected the length resource limits/requests to be 2 (memory & cpu), therefore the test failed since ephemeral memory is now also requested, changing the length to be 3 (memory, cpu and ephemeral storage).

It wasn't my intention to change the goal of the test which has nothing to do with ephemeral storage but with the explicitly "user-defined" resources. So this change only caused the test to more accurate.

Another test dedicated for ephemeral storage is also added.

File: pkg/container-disk/container-disk_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/container-disk/container-disk_test.go:197]: I agree, the question is where to defined this constant. 50M is needed both for compute container (which is defined in template.go) and for container disk (which is defined in container-disk.go).

I can define a different constant for each one since they don't have to be equal in their size.

File: pkg/virt-controller/services/template_test.go (hunk 11 above)
Comment (iholder-redhat) [reply pkg/virt-controller/services/template_test.go:2586]: As I wrote to vladik above: the thing is that if the user requests X MBs of ephemeral-storage then the amount that should be requested is X+50Mb since we should guarantee the user can use all of his requested storage.

File: pkg/container-disk/container-disk.go (hunk 10 above)
Comment (iholder-redhat) [reply pkg/container-disk/container-disk.go:194]: That's a good question :)
Should we?

File: pkg/container-disk/container-disk_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/container-disk/container-disk_test.go:188]: I don't mind this test being as it is now. As long as we have another test that explicitly checks the ephemeral storage, and it seems that we do. so it looks good to me.

File: pkg/virt-controller/services/template_test.go (hunk 11 above)
Comment (vladikr) [reply pkg/virt-controller/services/template_test.go:2586]: @enp0s3 we need to introduce this request to both containers.

**Comment by vladikr** [comment]
@iholder-redhat did you try running the functional test locally?

**Comment by vladikr** [comment]
@iholder-redhat did you try running the functional test locally?

**Comment by iholder-redhat** [comment]
@vladikr I figured out that as part of the refactoring I've accidentally replaced between "m" and "M" argument for MustParse. As you mentioned above the two differ as one refers to milliCPU metric and the other to Megabytes. This is indeed very confusing.

I ran unit-tests and some functional tests locally and now everything passes.

**Review by enp0s3 — APPROVED** [review]
Thank you!
/lgtm
/approve

**Review by enp0s3 — APPROVED** [review]
/lgtm
